### PR TITLE
[node] Enable app-specific signing keys to be calculated from an externally-provided callback

### PR DESCRIPTION
**Pull request by cf19drofxots**
## Summary

In order for the Node to produce state-update commitments, it needs access to some signing keys.

There are two ways in which this is supported:

1. An extended private key string is provided to the Node at the "EXTENDED_PRIVATE_KEY_PATH" key of the store service that is passed as an argument to the Node. If no such value exists for this key, the Node produces an extended private key and sets it at this key. This extended private key is then used to generate a "public identifer" for the Node (the address by which the Node is known by). It is also used to generate private keys which are specific to AppInstances.

2. Instead of supplying a mnemonic, the Node operator supplies two other arguments:

- an extended public key which will serve as the "public identifier" of the Node, and will be used to generate signer addresses at AppInstance-specific derivation paths for signature verification in the protocols.
- a callback function that offers the generation of a signing key given a specific derivation path. This enables the consumer of the Node (i.e. wallets) to not reveal any mnemonics but provide the ability to sign state isolated to specific AppInstances.

 The `Node` package exposes a reference implementation of the second approach through a function named `generatePrivateKeyGeneratorAndXPubPair` which produces these 2 arguments given an extended private key.

Main function of interest:
https://github.com/counterfactual/monorepo/blob/efd462bdeebaff63d264e109207bf095da285de9/packages/node/src/private-keys-generator.ts#L68-L113

----

This PR addresses signing issue: https://github.com/counterfactual/monorepo/issues/1916

Note: This _does not_  fix a security issue where currently on master, apps sharing the same appSequenceNumber across all channels use the same signing key. The fix is outside the scope of this PR.

### CODE REVIEW

File: packages/node/src/signing-keys-generator.ts [hunk 1]
```diff
@@ -0,0 +1,67 @@
+import { Node } from "@counterfactual/types";
+import { Wallet } from "ethers";
+import { hashMessage } from "ethers/utils";
+import { Memoize } from "typescript-memoize";
+
+export class SigningKeysGenerator {
```
Comment (snario) [inline packages/node/src/signing-keys-generator.ts]: Maybe a more explicit name is `PrivateKeyGenerator`?

File: packages/node/src/signing-keys-generator.ts [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+  @Memoize()
+  /**
+   * @param appSequenceNumber
+   * @param multisigAddress The namespace under which `appSequenceNumber` is
+   *        used since the provided callback function is idempotent for any
+   *        value passed in. This prevents collisions that'd occur on  using the
+   *        same `appSequenceNumber` across different channels from receiving the
+   *        same signing key from the callback.
+   */
+  public async getSigningKey(
```
Comment (snario) [inline packages/node/src/signing-keys-generator.ts]: `getPrivateKey`?

File: packages/node/src/signing-keys-generator.ts [hunk 3]
```diff
@@ -0,0 +1,67 @@
...
+      )!;
+    }
+
+    // FIXME: fallback on mnemonic-based signer if callback isn't specified
+    // in which case the derivation path needs to be converted into a decimal
+    // representation for it to be a valid path
+    const signingKey = await this.privateKeyGenerator(appInstanceUniqueId);
+    try {
+      new Wallet(signingKey);
+    } catch (e) {
```
Comment (snario) [inline packages/node/src/signing-keys-generator.ts]: I'm guessing `ethers` throws some very specific types of errors here, maybe catch those?

File: packages/node/src/private-keys-generator.ts [hunk 4]
```diff
@@ -0,0 +1,154 @@
...
+import { BigNumber } from "ethers/utils";
+import { fromExtendedKey, fromMnemonic } from "ethers/utils/hdnode";
+import log from "loglevel";
+import { Memoize } from "typescript-memoize";
+
+import { CF_PATH } from "./constants";
+
+export const EXTENDED_PRIVATE_KEY_PATH = "EXTENDED_PRIVATE_KEY";
+
+export class PrivateKeysGenerator {
```
Comment (IIIIllllIIIIllllIIIIllllIIIIllllIIIIll) [inline packages/node/src/private-keys-generator.ts]: `implements IPrivateKeyGenerator`?

File: packages/node/src/private-keys-generator.ts [hunk 5]
```diff
@@ -0,0 +1,154 @@
...
+    extendedPrvKey = fromMnemonic(Wallet.createRandom().mnemonic).extendedKey;
+    await storeService.set([
+      { key: EXTENDED_PRIVATE_KEY_PATH, value: extendedPrvKey }
+    ]);
+  } else {
+    log.info("Using extended private key found in the store.");
+  }
+  const [
+    privKeyGenerator,
+    pubExtendedKey
```
Comment (IIIIllllIIIIllllIIIIllllIIIIllllIIIIll) [inline packages/node/src/private-keys-generator.ts]: `extendedPubKey` (or `xpub`) for consistency

File: packages/node/src/node.ts [hunk 6]
```diff
@@ -59,9 +62,22 @@ export class Node {
     nodeConfig: NodeConfig,
     provider: BaseProvider,
     networkOrNetworkContext: "ropsten" | "kovan" | "rinkeby" | NetworkContext,
+    publicExtendedKey?: string,
+    privateKeyGenerator?: NodeTypes.IPrivateKeyGenerator,
     blocksNeededForConfirmation?: number
   ): Promise<Node> {
+    const [
+      privateKeysGenerator,
+      pubExtendedKey
```
Comment (IIIIllllIIIIllllIIIIllllIIIIllllIIIIll) [inline packages/node/src/node.ts]: `extendedPubKey` or `xpub`

File: packages/node/src/private-keys-generator.ts (hunk 4 above)
Comment (cf19drofxots) [reply packages/node/src/private-keys-generator.ts]: This class itself doesn't implement the interface, it's just a wrapper invoking the function that implements the interface and does the needed error checking for whether the generated private key is valid.

Comment (cf19drofxots) [reply packages/node/src/private-keys-generator.ts]: I'll rename the class, this is incorrectly named.

**Comment by rhlsthrm** [comment]
This has bumped up in priority for us. Can we get an update on if there's anything blocking this? Besides of course the master update.

**Pull request merged**